In [1]:
import pandas as pd
import numpy as np
import os

# Importing datasets

In [2]:
path = r'C:\Users\Aricia\MM-2020 Instacart Basket Analysis'

In [3]:
# Importing Customer data set
df_csts = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
# Investigating using describe
df_csts.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [5]:
# Counting rows and columns
df_csts.shape

(206209, 10)

In [6]:
# Checking columns 
df_csts.head(5)

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## Renaming columns

In [7]:
df_csts.rename(columns = {'Surnam' : 'Surname'}, inplace = True)

In [8]:
df_csts.rename(columns = {'STATE' : 'State'}, inplace = True)

In [9]:
df_csts.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)

In [10]:
df_csts.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [11]:
# Re-checking columns 
df_csts.head(5)

user_id First Name   Surname  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

Renamed all columns that had mispelling or title was not clear enough

## Dropping columns

In [12]:
# Dropping first name and surname columns as we already have user_id, which would be more useful for the analysis.
df_csts = df_csts.drop(columns = ['First Name', 'Surname'])

In [13]:
df_csts.shape

(206209, 8)

In [14]:
df_csts.head(5)

user_id  Gender       State  Age date_joined  number_of_dependants  \
0    26711  Female    Missouri   48    1/1/2017                     3   
1    33890  Female  New Mexico   36    1/1/2017                     0   
2    65803    Male       Idaho   35    1/1/2017                     2   
3   125935  Female        Iowa   40    1/1/2017                     0   
4   130797  Female    Maryland   26    1/1/2017                     1   

  marital_status  income  
0        married  165665  
1         single   59285  
2        married   99568  
3         single   42049  
4        married   40374

## Missing values

In [15]:
# Finding Missing Values 
df_csts.isnull().sum()

user_id                 0
Gender                  0
State                   0
Age                     0
date_joined             0
number_of_dependants    0
marital_status          0
income                  0
dtype: int64

No missing values

## Duplicates

In [16]:
# Finding Duplicates
df_dups_csts = df_csts[df_csts.duplicated()]

In [17]:
df_dups_csts

Empty DataFrame
Columns: [user_id, Gender, State, Age, date_joined, number_of_dependants, marital_status, income]
Index: []

No duplicates

## Mixed values

In [18]:
# Check for mixed types
for col in df_csts.columns.tolist():
  weird = (df_csts[[col]].map(type) != df_csts[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_csts[weird]) > 0:
    print (col)

No mixed data

## Combine prepared Instacart data

In [19]:
# Importing new dataset - optimized due to memory
df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_optimized.pkl'))

In [20]:
# Check the output
df_ords_prods.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day  busiest_days busiest_period_of_day  \
0  ...  Mid-range product  Regularly busy  Regular days           Most orders   
1  ...  Mid-range product  Regularly busy  Regular days        Average orders   
2  ...  Mid-range product     Busiest day  Busiest days        Average orders   
3  ...  Mid-range product  Regularly busy  Slowest days        Average orders   
4  ...  Mid-range product      Least busy  Slowest days        Average orders   

  max_order      loyalty_flag average_price spender_type order_frequency  \
0        32  Regular customer      6.935811  Low spender             8.0   
1        32  Regular customer      6.935811  Low spender             8.0   
2         5      New customer      7.930208  Low spender             6.0   
3         3      New customer      4.972414  Low spender             9.0   
4         3      New customer      4.972414  Low spender             9.0   

   order_frequency_flag  
0     Frequent customer  
1     Frequent customer  
2     Frequent customer  
3     Frequent customer  
4     Frequent customer  

[5 rows x 25 columns]

In [21]:
# Checking shape of new data
df_ords_prods.shape

(32434212, 25)

In [22]:
# Checking columns name
df_ords_prods.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'Merge_flag', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'average_price', 'spender_type', 'order_frequency',
       'order_frequency_flag'],
      dtype='object')

### Merging dataframes

Solving memory issues: MemoryError: Unable to allocate 247. MiB for an array with shape (32434212, 1) and data type object

In [23]:
# List current string/object flag columns
object_cols = ['_merge', 'Merge_flag', 'loyalty_flag', 'spender_type', 'order_frequency_flag']

In [24]:
for col in object_cols:
    if col in df_ords_prods.columns:
        df_ords_prods[col] = df_ords_prods[col].astype('category')

In [25]:
# Create a new dataframe combining both: df_ords_prods_csts_merged
df_ords_prods_csts_merged = df_csts.merge(df_ords_prods, on = 'user_id', indicator = 'Merge_flag49')

In [26]:
# Check for full match
df_ords_prods_csts_merged.head()

user_id  Gender     State  Age date_joined  number_of_dependants  \
0    26711  Female  Missouri   48    1/1/2017                     3   
1    26711  Female  Missouri   48    1/1/2017                     3   
2    26711  Female  Missouri   48    1/1/2017                     3   
3    26711  Female  Missouri   48    1/1/2017                     3   
4    26711  Female  Missouri   48    1/1/2017                     3   

  marital_status  income  product_id product_name  ...     busiest_day  \
0        married  165665         196         Soda  ...  Regularly busy   
1        married  165665         196         Soda  ...  Regularly busy   
2        married  165665         196         Soda  ...  Regularly busy   
3        married  165665        6184  Clementines  ...  Regularly busy   
4        married  165665        6184  Clementines  ...  Regularly busy   

   busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0  Busiest days         Average orders          8  New customer   
1  Regular days         Average orders          8  New customer   
2  Busiest days         Average orders          8  New customer   
3  Regular days         Average orders          8  New customer   
4  Slowest days            Most orders          8  New customer   

   average_price  spender_type  order_frequency  order_frequency_flag  \
0       7.988889   Low spender             19.0      Regular customer   
1       7.988889   Low spender             19.0      Regular customer   
2       7.988889   Low spender             19.0      Regular customer   
3       7.988889   Low spender             19.0      Regular customer   
4       7.988889   Low spender             19.0      Regular customer   

   Merge_flag49  
0          both  
1          both  
2          both  
3          both  
4          both  

[5 rows x 33 columns]

In [27]:
# Check for value counts
df_ords_prods_csts_merged['_merge'].value_counts()

_merge
both          32434212
left_only            0
right_only           0
Name: count, dtype: int64

### Exporting Data

In [28]:
# Export data to pkl
df_ords_prods_csts_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_ords_prods_csts_merged.pkl'))